test/mocha-junit-reporter-spec.js

```diff
@@ -120,411 +173,426 @@ describe('mocha-junit-reporter', function() {
+      var reporter = createReporter({attachments: true, outputs: true});
+      var test = createTest('has attachment');
       var filePath = '/path/to/file';
-      var testsuites;
-      var xml;
-      runner.startSuite(suite);
       test.attachments = [filePath];
       test.consoleOutputs = [ 'first console line', 'second console line' ];
-      runner.pass(test);
-      reporter.flush = function(suites) {
-        testsuites = suites;
-      };
-      runner.end();
-      expect(testsuites[0].testsuite[0]._attr.name).to.equal(suite.title);
-      expect(testsuites[0].testsuite[1].testcase).to.have.length(2);
-      expect(testsuites[0].testsuite[1].testcase[0]._attr.name).to.equal(test.fullTitle());
-      expect(testsuites[0].testsuite[1].testcase[1]).to.have.property('system-out', 'first console line\nsecond console line\n[[ATTACHMENT|' + filePath + ']]');
-      xml = reporter.getXml(testsuites);
-
-      expect(xml).to.include('<system-out>first console line\nsecond console line\n[[ATTACHMENT|' + filePath + ']]</system-out>');
-    });
-  });
 
-  describe('Output', function() {
-    var reporter, testsuites;
+      var suite = Suite.create(reporter.runner.suite, 'with attachments and outputs');
+      suite.addTest(test);
 
-    beforeEach(function() {
-      reporter = spyingReporter();
-    });
+      reporter.runner.run(function() {
+        expect(reporter._testsuites[1].testsuite[0]._attr.name).to.equal(suite.title);
+        expect(reporter._testsuites[1].testsuite[1].testcase).to.have.length(2);
+        expect(reporter._testsuites[1].testsuite[1].testcase[0]._attr.name).to.equal(test.fullTitle());
+        expect(reporter._testsuites[1].testsuite[1].testcase[1]).to.have.property('system-out', 'first console line\nsecond console line\n[[ATTACHMENT|' + filePath + ']]');
 
-    it('skips suites with empty title', function() {
-      runner.startSuite({title: '', tests: [1]});
-      runner.end();
+        expect(reporter._xml).to.include('<system-out>first console line\nsecond console line\n[[ATTACHMENT|' + filePath + ']]</system-out>');
 
-      expect(testsuites).to.be.empty;
+        done();
+      });
     });
+  });
 
-    it('skips suites without testcases and suites', function() {
-      runner.startSuite({title: 'test me'});
-      runner.end();
-
-      expect(testsuites).to.be.empty;
+  describe('Output', function() {
+    it('skips suites with empty title', function(done) {
+      var reporter = createReporter();
+      var suite = Suite.create(reporter.runner.suite, '');
+      suite.root = false; // mocha treats suites with empty title as root, so not sure this is possible
+      suite.addTest(createTest('test'));
+
+      reporter.runner.run(function() {
+        expect(reporter._testsuites).to.have.lengthOf(1);
+        expect(reporter._testsuites[0].testsuite[0]._attr.name).to.equal('Root Suite');
+        done();
+      });
     });
 
-    it('does not skip suites with nested suites', function() {
```

Comment:
This test didn't match Mocha's behavior. Mocha doesn't emit the `suite` event if a suite has no tests, even if it has empty nested suites. See https://github.com/mochajs/mocha/blob/c0137eb698add08f29035467ea1dc230904f82ba/lib/runner.js#L723.